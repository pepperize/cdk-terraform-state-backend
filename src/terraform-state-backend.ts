import { aws_dynamodb, aws_s3 } from "aws-cdk-lib";
import { Construct } from "constructs";

export interface TerraformStateBackendProps {
  readonly bucketName: string;
  readonly tableName: string;
}

export class TerraformStateBackend extends Construct {
  readonly bucket: aws_s3.IBucket;
  readonly table: aws_dynamodb.ITable;

  public constructor(scope: Construct, id: string, props: TerraformStateBackendProps) {
    super(scope, id);

    const { bucketName, tableName } = props;

    this.bucket = new aws_s3.Bucket(this, "Bucket", {
      bucketName: bucketName,
      versioned: true,
      encryption: aws_s3.BucketEncryption.KMS_MANAGED,
      blockPublicAccess: aws_s3.BlockPublicAccess.BLOCK_ALL,
    });

    this.table = new aws_dynamodb.Table(scope, "Table", {
      tableName: tableName,
      billingMode: aws_dynamodb.BillingMode.PAY_PER_REQUEST,
      partitionKey: {
        name: "LockID",
        type: aws_dynamodb.AttributeType.STRING,
      },
      encryption: aws_dynamodb.TableEncryption.DEFAULT,
      pointInTimeRecovery: true,
    });
  }
}
