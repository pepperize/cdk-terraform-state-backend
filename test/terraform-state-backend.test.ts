import { App, Aspects, Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { AwsSolutionsChecks } from "cdk-nag";
import { TerraformStateBackend } from "../src";

describe("TerraformStateBackend", () => {
  it("Should created bucket that is versioned, encrypted and blocks public access", () => {
    // Given
    const app = new App();
    const stack = new Stack(app, "stack", {
      env: {
        account: "0",
        region: "us-east-1",
      },
    });

    // When
    new TerraformStateBackend(stack, "TerraformStateBackend", {
      bucketName: "terraform-state-backend",
      tableName: "terraform-state-backend",
    });
    const template = Template.fromStack(stack);

    // Then
    template.hasResourceProperties("AWS::S3::Bucket", {
      VersioningConfiguration: {
        Status: "Enabled",
      },
      BucketEncryption: {
        ServerSideEncryptionConfiguration: [
          {
            ServerSideEncryptionByDefault: {
              SSEAlgorithm: "aws:kms",
            },
          },
        ],
      },
      PublicAccessBlockConfiguration: {
        BlockPublicAcls: true,
        BlockPublicPolicy: true,
        IgnorePublicAcls: true,
        RestrictPublicBuckets: true,
      },
    });
  });

  it("Should comply to best practices", () => {
    // Given
    const app = new App();
    const stack = new Stack(app, "stack", {
      env: {
        account: "0",
        region: "us-east-1",
      },
    });

    // When
    new TerraformStateBackend(stack, "TerraformStateBackend", {
      bucketName: "terraform-state-backend",
      tableName: "terraform-state-backend",
    });

    // Then
    Aspects.of(app).add(new AwsSolutionsChecks({ verbose: true }));
  });

  it("Should match snapshot", () => {
    // Given
    const app = new App();
    const stack = new Stack(app, "stack", {
      env: {
        account: "0",
        region: "us-east-1",
      },
    });

    // When
    new TerraformStateBackend(stack, "TerraformStateBackend", {
      bucketName: "terraform-state-backend",
      tableName: "terraform-state-backend",
    });
    const template = Template.fromStack(stack);

    // Then
    expect(template).toMatchSnapshot();
  });
});
