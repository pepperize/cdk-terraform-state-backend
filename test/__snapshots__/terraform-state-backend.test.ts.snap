// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TerraformStateBackend Should match snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "TableCD117FA1": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "LockID",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "LockID",
            "KeyType": "HASH",
          },
        ],
        "PointInTimeRecoverySpecification": Object {
          "PointInTimeRecoveryEnabled": true,
        },
        "SSESpecification": Object {
          "SSEEnabled": false,
        },
        "TableName": "terraform-state-backend",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "TerraformStateBackendBucket7CE90B83": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "BucketName": "terraform-state-backend",
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": Object {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
